libs/vkd3d/command.c

```diff
@@ -6344,6 +6350,42 @@ static void STDMETHODCALLTYPE d3d12_command_list_CopyTextureRegion(d3d12_command
         return;
 
     d3d12_command_list_ensure_batch(list, copy_info.batch_type);
+    alias = false;
+    format_match = true;
+    for (size_t i = 0; i < list->transfer_batch.ctr_batch_len; i++) {
```

Comment:
This section is hard to reason about. I see it's disabling the alias checking for barriers, but I'm not sure if that's what we want here.

COPY_DEST resource state in D3D12 is kinda broken. Overlapping copies to same subresource must be synchronized implicitly by implementation. So, we'd need to either break batch, or add an internal TRANSFER memory barrier between two copies that overlap to resolve the write-after-write hazard.

Reply:
Good catch, I'll remove the format_match condition.

As for why it has to be done here, it's because when there's an alias, we need to flush not only the barriers but also the copies and the post-barriers as well.

Reply:
Actually, on second thought, this is quite troubling for Forza. It has multiple copies to the same image that doesn't overlap. But to check that we need to compare the coordinates as well. It does fit within the quadratic implementation though.

An alternative is to try to use render-based blits, which is guaranteed to be ordered, but that's going to be another massively overengineered hell...

Reply:
Yes, exactly. I think I’ll just compromise here and go the less optimized path. It looks like the game developer is well aware of the implications as they interleave calls to different destination images; so while not having box overlap checks will make this suboptimal, it is still an improvement from having a barrier after each copy.